network/local.go

```diff
@@ -0,0 +1,459 @@
+	addr Address
+	// are we listening or not
+	listening bool
+
+	sync.Mutex
+
+	// quit is used to stop the listening routine
+	quit chan bool
+
+	ctx *LocalContext
+}
+
+// NewLocalListener returns a fresh LocalListener which implements the Listener
+// interface.
+func NewLocalListener(addr Address) (*LocalListener, error) {
+	return NewLocalListenerWithContext(defaultLocalContext, addr)
+}
+
+// NewLocalListenerWithContext is similar to NewLocalListener but taking a
+// specific LocalContext to use to communicate.
+func NewLocalListenerWithContext(ctx *LocalContext, addr Address) (*LocalListener, error) {
+	l := &LocalListener{
+		quit: make(chan bool),
+		ctx:  ctx,
+	}
+	return l, l.bind(addr)
+}
+
+func (ll *LocalListener) bind(addr Address) error {
+	ll.Lock()
+	defer ll.Unlock()
+	if addr.ConnType() != Local {
+		return errors.New("Wrong address type for local listener")
+	}
+	if ll.ctx.IsListening(addr) {
+		return fmt.Errorf("%s is already listening: can't listen again", addr)
+	}
+	ll.addr = addr
+	return nil
+}
+
+// Listen implements the Listener interface
+func (ll *LocalListener) Listen(fn func(Conn)) error {
+	ll.Lock()
+	ll.quit = make(chan bool)
+	ll.ctx.Listening(ll.addr, fn)
+	ll.listening = true
+	ll.Unlock()
+
+	<-ll.quit
+	return nil
+}
+
+// Stop implements the Listener interface
+func (ll *LocalListener) Stop() error {
+	ll.Lock()
+	defer ll.Unlock()
+	ll.ctx.StopListening(ll.addr)
+	if ll.listening {
+		close(ll.quit)
+	}
+	ll.listening = false
+	return nil
+}
+
+// Address returns the listening address used.
+func (ll *LocalListener) Address() Address {
+	ll.Lock()
+	defer ll.Unlock()
+	return ll.addr
+}
+
+// Listening returns true if this Listener is actually listening for any
+// incoming connections.
+func (ll *LocalListener) Listening() bool {
+	ll.Lock()
+	defer ll.Unlock()
+	return ll.listening
+}
+
+// LocalHost is a Host implementation using LocalConn and LocalListener as
+// the underlying means of communication. It is a implementation of the Host
+// interface.
+type LocalHost struct {
+	addr Address
+	*LocalListener
+	ctx *LocalContext
+}
+
+// NewLocalHost returns a fresh Host using Local communication that will listen
+// on the given addr.
+func NewLocalHost(addr Address) (*LocalHost, error) {
+	return NewLocalHostWithContext(defaultLocalContext, addr)
+}
+
+// NewLocalHostWithContext is similar to NewLocalHost but specify which
+// LocalContext to use for communicating.
+func NewLocalHostWithContext(ctx *LocalContext, addr Address) (*LocalHost, error) {
+	lh := &LocalHost{
+		addr: addr,
+		ctx:  ctx,
+	}
+	var err error
+	lh.LocalListener, err = NewLocalListenerWithContext(ctx, addr)
+	return lh, err
+
+}
+
+// Connect implements the Host interface.
+func (lh *LocalHost) Connect(addr Address) (Conn, error) {
+	if addr.ConnType() != Local {
+		return nil, errors.New("Can't connect to non-Local address")
+	}
+	var finalErr error
+	for i := 0; i < MaxRetryConnect; i++ {
```

Comment:
Not failure but it's the same story with TCP: If we launch the Listening part in a go routine, we don't know when the go routine is actually launched so sometimes it fails if we just try to connect once.